Hapi Dependency on Charm++

Currently, Hapi depends on the Charm++ runtime (for `QdCreate`, `QdProcess` and `CkCallback::send`), which means that non-Charm++ applications (such as those that only use Converse) now require the Charm++ runtime to build and run. Converse itself requires Hapi for `hapiPollEvents` and `exitHybridAPI`.

See also #2459.

This causes the current auto build failure with cuda:
```
../../bin/charmc -optimize -production   -o check check.o -language converse++
../../bin/../lib/libcudahybridapi.a(hybridapi.o): In function `hapiPollEvents':
hapi_impl.cpp:(.text+0x1590): undefined reference to `CkCallback::send(void*) const'
hapi_impl.cpp:(.text+0x169a): undefined reference to `QdProcess(int)'
../../bin/../lib/libcudahybridapi.a(hybridapi.o): In function `hapiAddCallback':
hapi_impl.cpp:(.text+0x1afb): undefined reference to `QdCreate(int)'
../../bin/../lib/libcudahybridapi.a(hybridapi.o): In function `hapiEnqueue':
hapi_impl.cpp:(.text+0x1c41): undefined reference to `QdCreate(int)'
hapi_impl.cpp:(.text+0x1d17): undefined reference to `QdCreate(int)'
hapi_impl.cpp:(.text+0x1e56): undefined reference to `QdCreate(int)'
```

--- a/src/arch/cuda/hybridAPI/hapi_impl.cpp
+++ b/src/arch/cuda/hybridAPI/hapi_impl.cpp
@@ -510,20 +510,12 @@ static void* hostToDeviceCallback(void* arg) {
   NVTXTracer nvtx_range("hostToDeviceCallback", NVTXColor::Asbestos);
 #endif
   hapiWorkRequest* wr = *((hapiWorkRequest**)((char*)arg + CmiMsgHeaderSizeBytes + sizeof(int)));
-  if (hapiInvokeCallback) {
-    hapiInvokeCallback(wr->host_to_device_cb);
-  }
-  else {
-    CmiAbort("hapiInvokeCallback is not set");
-  }
+  CmiAssert(hapiInvokeCallback);
+  hapiInvokeCallback(wr->host_to_device_cb);
 
   // inform QD that the host-to-device transfer is complete
-  if (hapiQdProcess) {
-    hapiQdProcess(1);
-  }
-  else {
-    CmiAbort("hapiQdProcess is not set");
-  }
+  CmiAssert(hapiQdProcess);
+  hapiQdProcess(1);
 
   return NULL;
 }
@@ -534,20 +526,12 @@ static void* kernelCallback(void* arg) {
   NVTXTracer nvtx_range("kernelCallback", NVTXColor::Asbestos);
 #endif
   hapiWorkRequest* wr = *((hapiWorkRequest**)((char*)arg + CmiMsgHeaderSizeBytes + sizeof(int)));
-  if (hapiInvokeCallback) {
-    hapiInvokeCallback(wr->kernel_cb);
-  }
-  else {
-    CmiAbort("hapiInvokeCallback is not set");
-  }
+  CmiAssert(hapiInvokeCallback);
+  hapiInvokeCallback(wr->kernel_cb);
 
   // inform QD that the kernel is complete
-  if (hapiQdProcess) {
-    hapiQdProcess(1);
-  }
-  else {
-    CmiAbort("hapiQdProcess is not set");
-  }
+  CmiAssert(hapiQdProcess);
+  hapiQdProcess(1);
 
   return NULL;
 }
@@ -562,23 +546,15 @@ static void* deviceToHostCallback(void* arg) {
 
   // invoke user callback
   if (wr->device_to_host_cb) {
-    if (hapiInvokeCallback) {
-      hapiInvokeCallback(wr->device_to_host_cb);
-    }
-    else {
-      CmiAbort("hapiInvokeCallback is not set");
-    }
+    CmiAssert(hapiInvokeCallback);
+    hapiInvokeCallback(wr->device_to_host_cb);
   }
 
   hapiWorkRequestCleanup(wr);
 
   // inform QD that device-to-host transfer is complete
-  if (hapiQdProcess) {
-    hapiQdProcess(1);
-  }
-  else {
-    CmiAbort("hapiQdProcess is not set");
-  }
+  CmiAssert(hapiQdProcess);
+  hapiQdProcess(1);
 
   return NULL;
 }
@@ -594,21 +570,13 @@ static void* lightCallback(void *arg) {
 
   // invoke user callback
   if (cb) {
-    if (hapiInvokeCallback) {
-      hapiInvokeCallback(cb);
-    }
-    else {
-      CmiAbort("hapiInvokeCallback is not set");
-    }
+    CmiAssert(hapiInvokeCallback);
+    hapiInvokeCallback(cb);
   }
 
   // notify process to QD
-  if (hapiQdProcess) {
-    hapiQdProcess(1);
-  }
-  else {
-    CmiAbort("hapiQdProcess is not set");
-  }
+  CmiAssert(hapiQdProcess);
+  hapiQdProcess(1);
 
   return NULL;
 }
@@ -710,12 +678,8 @@ void hapiEnqueue(hapiWorkRequest* wr) {
   if (wr->host_to_device_cb) {
     // while there is an ongoing workrequest, quiescence should not be detected
     // even if all PEs seem idle
-    if (hapiQdCreate) {
-      hapiQdCreate(1);
-    }
-    else {
-      CmiAbort("hapiQdCreate is not set");
-    }
+    CmiAssert(hapiQdCreate);
+    hapiQdCreate(1);
 
 #ifdef HAPI_CUDA_CALLBACK
     addCallback(wr, AfterHostToDevice);
@@ -729,12 +693,8 @@ void hapiEnqueue(hapiWorkRequest* wr) {
 
   // add kernel callback
   if (wr->kernel_cb) {
-    if (hapiQdCreate) {
-      hapiQdCreate(1);
-    }
-    else {
-      CmiAbort("hapiQdCreate is not set");
-    }
+    CmiAssert(hapiQdCreate);
+    hapiQdCreate(1);
 
 #ifdef HAPI_CUDA_CALLBACK
     addCallback(wr, AfterKernel);
@@ -747,12 +707,8 @@ void hapiEnqueue(hapiWorkRequest* wr) {
   CsvAccess(gpu_manager).deviceToHostTransfer(wr);
 
   // add device-to-host transfer callback
-  if (hapiQdCreate) {
-    hapiQdCreate(1);
-  }
-  else {
-    CmiAbort("hapiQdCreate is not set");
-  }
+  CmiAssert(hapiQdCreate);
+  hapiQdCreate(1);
 #ifdef HAPI_CUDA_CALLBACK
   // always invoked to free memory
   addCallback(wr, AfterDeviceToHost);
@@ -1257,12 +1213,8 @@ void hapiPollEvents() {
     if (cudaEventQuery(hev.event) == cudaSuccess) {
       // invoke Charm++ callback if one was given
       if (hev.cb) {
-        if (hapiInvokeCallback) {
-          hapiInvokeCallback(hev.cb, hev.cb_msg);
-        }
-        else {
-          CmiAbort("hapiInvokeCallback is not set");
-        }
+        CmiAssert(hapiInvokeCallback);
+        hapiInvokeCallback(hev.cb, hev.cb_msg);
       }
 
       // clean up hapiWorkRequest
@@ -1274,12 +1226,8 @@ void hapiPollEvents() {
       CpvAccess(n_hapi_events)--;
 
       // inform QD that an event was processed
-      if (hapiQdProcess) {
-        hapiQdProcess(1);
-      }
-      else {
-        CmiAbort("hapiQdProcess is not set");
-      }
+      CmiAssert(hapiQdProcess);
+      hapiQdProcess(1);
     }
     else {
       // stop going through the queue once we encounter a non-successful event
@@ -1350,12 +1298,8 @@ void hapiAddCallback(cudaStream_t stream, void* cb, void* cb_msg) {
 
   // while there is an ongoing workrequest, quiescence should not be detected
   // even if all PEs seem idle
-  if (hapiQdCreate) {
-    hapiQdCreate(1);
-  }
-  else {
-    CmiAbort("hapiQdCreate is not set");
-  }
+  CmiAssert(hapiQdCreate);
+  hapiQdCreate(1);
 }
 
 cudaError_t hapiMalloc(void** devPtr, size_t size) {

--- a/src/arch/cuda/hybridAPI/hapi_impl.cpp
+++ b/src/arch/cuda/hybridAPI/hapi_impl.cpp
@@ -32,10 +32,6 @@ typedef struct gpuEventTimer {
 } gpuEventTimer;
 #endif
 
-#ifdef HAPI_INSTRUMENT_WRS
-static bool initializedInstrument();
-#endif
-
 // Event stages used for profiling.
 enum WorkRequestStage{
   DataSetup        = 1,
@@ -81,11 +77,11 @@ static inline int getMyCudaDevice(int my_node) {
 }
 
 // Used to invoke user's Charm++ callback function
-void (*hapiInvokeCallback)(void*, void*);
+void (*hapiInvokeCallback)(void*, void*) = NULL;
 
 // Functions used to support quiescence detection.
-void (*hapiQdCreate)(int);
-void (*hapiQdProcess)(int);
+void (*hapiQdCreate)(int) = NULL;
+void (*hapiQdProcess)(int) = NULL;
 
 // Initial size of the user-addressed portion of host/device buffer arrays;
 // the system-addressed portion of host/device buffer arrays (used when there
@@ -146,7 +142,7 @@ public:
 #endif
 
 #ifdef HAPI_INSTRUMENT_WRS
-  std::vector<std::vector<std::vector<hapiRequestTimeInfo> > > avg_times_;
+  std::vector<std::vector<std::vector<hapiRequestTimeInfo>>> avg_times_;
   bool init_instr_;
 #endif
 
@@ -155,6 +151,7 @@ public:
   CmiNodeLock progress_lock_;
   CmiNodeLock stream_lock_;
   CmiNodeLock mempool_lock_;
+  CmiNodeLock inst_lock_;
 #endif
 
   cudaDeviceProp device_prop_;
@@ -194,6 +191,7 @@ void GPUManager::init() {
   progress_lock_ = CmiCreateLock();
   stream_lock_ = CmiCreateLock();
   mempool_lock_ = CmiCreateLock();
+  inst_lock_ = CmiCreateLock();
 #endif
 
 #ifdef HAPI_TRACE
@@ -512,10 +510,20 @@ static void* hostToDeviceCallback(void* arg) {
   NVTXTracer nvtx_range("hostToDeviceCallback", NVTXColor::Asbestos);
 #endif
   hapiWorkRequest* wr = *((hapiWorkRequest**)((char*)arg + CmiMsgHeaderSizeBytes + sizeof(int)));
-  hapiInvokeCallback(wr->host_to_device_cb);
+  if (hapiInvokeCallback) {
+    hapiInvokeCallback(wr->host_to_device_cb);
+  }
+  else {
+    CmiAbort("hapiInvokeCallback is not set");
+  }
 
   // inform QD that the host-to-device transfer is complete
-  hapiQdProcess(1);
+  if (hapiQdProcess) {
+    hapiQdProcess(1);
+  }
+  else {
+    CmiAbort("hapiQdProcess is not set");
+  }
 
   return NULL;
 }
@@ -526,10 +534,20 @@ static void* kernelCallback(void* arg) {
   NVTXTracer nvtx_range("kernelCallback", NVTXColor::Asbestos);
 #endif
   hapiWorkRequest* wr = *((hapiWorkRequest**)((char*)arg + CmiMsgHeaderSizeBytes + sizeof(int)));
-  hapiInvokeCallback(wr->kernel_cb);
+  if (hapiInvokeCallback) {
+    hapiInvokeCallback(wr->kernel_cb);
+  }
+  else {
+    CmiAbort("hapiInvokeCallback is not set");
+  }
 
   // inform QD that the kernel is complete
-  hapiQdProcess(1);
+  if (hapiQdProcess) {
+    hapiQdProcess(1);
+  }
+  else {
+    CmiAbort("hapiQdProcess is not set");
+  }
 
   return NULL;
 }
@@ -544,13 +562,23 @@ static void* deviceToHostCallback(void* arg) {
 
   // invoke user callback
   if (wr->device_to_host_cb) {
-    hapiInvokeCallback(wr->device_to_host_cb);
+    if (hapiInvokeCallback) {
+      hapiInvokeCallback(wr->device_to_host_cb);
+    }
+    else {
+      CmiAbort("hapiInvokeCallback is not set");
+    }
   }
 
   hapiWorkRequestCleanup(wr);
 
   // inform QD that device-to-host transfer is complete
-  hapiQdProcess(1);
+  if (hapiQdProcess) {
+    hapiQdProcess(1);
+  }
+  else {
+    CmiAbort("hapiQdProcess is not set");
+  }
 
   return NULL;
 }
@@ -566,11 +594,21 @@ static void* lightCallback(void *arg) {
 
   // invoke user callback
   if (cb) {
-    hapiInvokeCallback(cb);
+    if (hapiInvokeCallback) {
+      hapiInvokeCallback(cb);
+    }
+    else {
+      CmiAbort("hapiInvokeCallback is not set");
+    }
   }
 
   // notify process to QD
-  hapiQdProcess(1);
+  if (hapiQdProcess) {
+    hapiQdProcess(1);
+  }
+  else {
+    CmiAbort("hapiQdProcess is not set");
+  }
 
   return NULL;
 }
@@ -672,7 +710,12 @@ void hapiEnqueue(hapiWorkRequest* wr) {
   if (wr->host_to_device_cb) {
     // while there is an ongoing workrequest, quiescence should not be detected
     // even if all PEs seem idle
-    hapiQdCreate(1);
+    if (hapiQdCreate) {
+      hapiQdCreate(1);
+    }
+    else {
+      CmiAbort("hapiQdCreate is not set");
+    }
 
 #ifdef HAPI_CUDA_CALLBACK
     addCallback(wr, AfterHostToDevice);
@@ -686,7 +729,12 @@ void hapiEnqueue(hapiWorkRequest* wr) {
 
   // add kernel callback
   if (wr->kernel_cb) {
-    hapiQdCreate(1);
+    if (hapiQdCreate) {
+      hapiQdCreate(1);
+    }
+    else {
+      CmiAbort("hapiQdCreate is not set");
+    }
 
 #ifdef HAPI_CUDA_CALLBACK
     addCallback(wr, AfterKernel);
@@ -699,7 +747,12 @@ void hapiEnqueue(hapiWorkRequest* wr) {
   CsvAccess(gpu_manager).deviceToHostTransfer(wr);
 
   // add device-to-host transfer callback
-  hapiQdCreate(1);
+  if (hapiQdCreate) {
+    hapiQdCreate(1);
+  }
+  else {
+    CmiAbort("hapiQdCreate is not set");
+  }
 #ifdef HAPI_CUDA_CALLBACK
   // always invoked to free memory
   addCallback(wr, AfterDeviceToHost);
@@ -858,16 +911,19 @@ static inline void hapiWorkRequestStartTime(hapiWorkRequest* wr) {
 static inline void profileWorkRequestEvent(hapiWorkRequest* wr,
                                            WorkRequestStage event) {
 #ifdef HAPI_INSTRUMENT_WRS
-  if (initializedInstrument()) {
+#if CMK_SMP || CMK_MULTICORE
+  CmiLock(CsvAccess(gpu_manager).inst_lock_);
+#endif
+
+  if (CsvAccess(gpu_manager).init_instr_) {
     double tt = CmiWallTimer() - (wr->phase_start_time);
     int index = wr->chare_index;
     char type = wr->comp_type;
     char phase = wr->comp_phase;
 
-    std::vector<hapiRequestTimeInfo> &vec = wr->avg_times_[index][type];
-    if (vec.size() <= phase){
-      vec.growAtLeast(phase);
-      vec.size() = phase+1;
+    std::vector<hapiRequestTimeInfo> &vec = CsvAccess(gpu_manager).avg_times_[index][type];
+    if (vec.size() <= phase) {
+      vec.resize(phase+1);
     }
     switch (event) {
       case DataSetup:
@@ -884,7 +940,14 @@ static inline void profileWorkRequestEvent(hapiWorkRequest* wr,
         printf("[HAPI] invalid event during profileWorkRequestEvent\n");
     }
   }
+  else {
+    printf("[HAPI] instrumentation not initialized!\n");
+  }
+
+#if CMK_SMP || CMK_MULTICORE
+  CmiUnlock(CsvAccess(gpu_manager).inst_lock_);
 #endif
+#endif // HAPI_INSTRUMENT_WRS
 }
 
 // Create a pool with n_slots slots.
@@ -1127,44 +1190,58 @@ void hapiPoolFree(void* ptr) {
 }
 
 #ifdef HAPI_INSTRUMENT_WRS
-void hapiInitInstrument(int n_chares, char n_types) {
-  avg_times_.reserve(n_chares);
-  avg_times_.size() = n_chares;
-  for (int i = 0; i < n_chares; i++) {
-    avg_times_[i].reserve(n_types);
-    avg_times_[i].size() = n_types;
-  }
-  init_instr_ = true;
-}
+void hapiInitInstrument(int n_chares, int n_types) {
+#if CMK_SMP || CMK_MULTICORE
+  CmiLock(CsvAccess(gpu_manager).inst_lock_);
+#endif
 
-static bool initializedInstrument() {
-  return init_instr_;
+  if (!CsvAccess(gpu_manager).init_instr_) {
+    CsvAccess(gpu_manager).avg_times_.resize(n_chares);
+    for (int i = 0; i < n_chares; i++) {
+      CsvAccess(gpu_manager).avg_times_[i].resize(n_types);
+    }
+    CsvAccess(gpu_manager).init_instr_ = true;
+  }
+
+#if CMK_SMP || CMK_MULTICORE
+  CmiUnlock(CsvAccess(gpu_manager).inst_lock_);
+#endif
 }
 
 hapiRequestTimeInfo* hapiQueryInstrument(int chare, char type, char phase) {
-  if (phase < avg_times_[chare][type].size()) {
-    return &avg_times_[chare][type][phase];
+#if CMK_SMP || CMK_MULTICORE
+  CmiLock(CsvAccess(gpu_manager).inst_lock_);
+#endif
+
+  if (phase < CsvAccess(gpu_manager).avg_times_[chare][type].size()) {
+    return &CsvAccess(gpu_manager).avg_times_[chare][type][phase];
   }
   else {
     return NULL;
   }
+
+#if CMK_SMP || CMK_MULTICORE
+  CmiUnlock(CsvAccess(gpu_manager).inst_lock_);
+#endif
 }
 
 void hapiClearInstrument() {
-  for (int chare = 0; chare < avg_times_.size(); chare++) {
-    for (int type = 0; type < avg_times_[chare].size(); type++) {
-      for (int phase = 0; phase < avg_times_[chare][type].size(); phase++) {
-        avg_times_[chare][type][phase].transferTime = 0.0;
-        avg_times_[chare][type][phase].kernelTime = 0.0;
-        avg_times_[chare][type][phase].cleanupTime = 0.0;
-        avg_times_[chare][type][phase].n = 0;
-      }
-      avg_times_[chare][type].size() = 0;
+#if CMK_SMP || CMK_MULTICORE
+  CmiLock(CsvAccess(gpu_manager).inst_lock_);
+#endif
+
+  for (int chare = 0; chare < CsvAccess(gpu_manager).avg_times_.size(); chare++) {
+    for (char type = 0; type < CsvAccess(gpu_manager).avg_times_[chare].size(); type++) {
+      CsvAccess(gpu_manager).avg_times_[chare][type].clear();
     }
-    avg_times_[chare].size() = 0;
+    CsvAccess(gpu_manager).avg_times_[chare].clear();
   }
-  avg_times_.size() = 0;
-  init_instr_ = false;
+  CsvAccess(gpu_manager).avg_times_.clear();
+  CsvAccess(gpu_manager).init_instr_ = false;
+
+#if CMK_SMP || CMK_MULTICORE
+  CmiUnlock(CsvAccess(gpu_manager).inst_lock_);
+#endif
 }
 #endif // HAPI_INSTRUMENT_WRS
 
@@ -1180,7 +1257,12 @@ void hapiPollEvents() {
     if (cudaEventQuery(hev.event) == cudaSuccess) {
       // invoke Charm++ callback if one was given
       if (hev.cb) {
-        hapiInvokeCallback(hev.cb, hev.cb_msg);
+        if (hapiInvokeCallback) {
+          hapiInvokeCallback(hev.cb, hev.cb_msg);
+        }
+        else {
+          CmiAbort("hapiInvokeCallback is not set");
+        }
       }
 
       // clean up hapiWorkRequest
@@ -1192,7 +1274,12 @@ void hapiPollEvents() {
       CpvAccess(n_hapi_events)--;
 
       // inform QD that an event was processed
-      hapiQdProcess(1);
+      if (hapiQdProcess) {
+        hapiQdProcess(1);
+      }
+      else {
+        CmiAbort("hapiQdProcess is not set");
+      }
     }
     else {
       // stop going through the queue once we encounter a non-successful event
@@ -1263,7 +1350,12 @@ void hapiAddCallback(cudaStream_t stream, void* cb, void* cb_msg) {
 
   // while there is an ongoing workrequest, quiescence should not be detected
   // even if all PEs seem idle
-  hapiQdCreate(1);
+  if (hapiQdCreate) {
+    hapiQdCreate(1);
+  }
+  else {
+    CmiAbort("hapiQdCreate is not set");
+  }
 }
 
 cudaError_t hapiMalloc(void** devPtr, size_t size) {

--- a/src/arch/cuda/hybridAPI/hapi_impl.cpp
+++ b/src/arch/cuda/hybridAPI/hapi_impl.cpp
@@ -1,15 +1,15 @@
-#include "hapi.h"
-#include "hapi_impl.h"
-#include "converse.h"
-#include "ckcallback.h"
-#include "cklists.h"
 #include <stdio.h>
 #include <stdlib.h>
 #include <algorithm>
 #include <queue>
 #include <atomic>
-#include "cuda.h"
+#include <vector>
 
+#include <cuda_runtime.h>
+
+#include "converse.h"
+#include "hapi.h"
+#include "hapi_impl.h"
 #ifdef HAPI_NVTX_PROFILE
 #include "hapi_nvtx.h"
 #endif
@@ -80,14 +80,12 @@ static inline int getMyCudaDevice(int my_node) {
   return my_node % device_count;
 }
 
-// A function in ck.C which casts the void* to a CkCallback object and invokes
-// the Charm++ callback.
-extern void CUDACallbackManager(void* fn);
-extern int CmiMyPe();
+// Used to invoke user's Charm++ callback function
+void (*hapiInvokeCallback)(void*, void*);
 
 // Functions used to support quiescence detection.
-extern void QdCreate(int n);
-extern void QdProcess(int n);
+void (*hapiQdCreate)(int);
+void (*hapiQdProcess)(int);
 
 // Initial size of the user-addressed portion of host/device buffer arrays;
 // the system-addressed portion of host/device buffer arrays (used when there
@@ -113,8 +111,8 @@ public:
 // Scale the amount of memory each node pins.
 #define HAPI_MEMPOOL_SCALE 1.0
 
-  CkVec<BufferPool> mempool_free_bufs_;
-  CkVec<size_t> mempool_boundaries_;
+  std::vector<BufferPool> mempool_free_bufs_;
+  std::vector<size_t> mempool_boundaries_;
   bool mempool_initialized_;
 
   // The runtime system keeps track of all allocated buffers on the GPU.
@@ -148,7 +146,7 @@ public:
 #endif
 
 #ifdef HAPI_INSTRUMENT_WRS
-  CkVec<CkVec<CkVec<hapiRequestTimeInfo> > > avg_times_;
+  std::vector<std::vector<std::vector<hapiRequestTimeInfo> > > avg_times_;
   bool init_instr_;
 #endif
 
@@ -235,7 +233,6 @@ void GPUManager::init() {
   // set up mempool metadata
   mempool_initialized_ = false;
   mempool_boundaries_.reserve(HAPI_MEMPOOL_NUM_SLOTS);
-  mempool_boundaries_.length() = HAPI_MEMPOOL_NUM_SLOTS;
 
   size_t buf_size = HAPI_MEMPOOL_MIN_BUFFER_SIZE;
   for(int i = 0; i < HAPI_MEMPOOL_NUM_SLOTS; i++){
@@ -515,10 +512,10 @@ static void* hostToDeviceCallback(void* arg) {
   NVTXTracer nvtx_range("hostToDeviceCallback", NVTXColor::Asbestos);
 #endif
   hapiWorkRequest* wr = *((hapiWorkRequest**)((char*)arg + CmiMsgHeaderSizeBytes + sizeof(int)));
-  CUDACallbackManager(wr->host_to_device_cb);
+  hapiInvokeCallback(wr->host_to_device_cb);
 
   // inform QD that the host-to-device transfer is complete
-  QdProcess(1);
+  hapiQdProcess(1);
 
   return NULL;
 }
@@ -529,10 +526,10 @@ static void* kernelCallback(void* arg) {
   NVTXTracer nvtx_range("kernelCallback", NVTXColor::Asbestos);
 #endif
   hapiWorkRequest* wr = *((hapiWorkRequest**)((char*)arg + CmiMsgHeaderSizeBytes + sizeof(int)));
-  CUDACallbackManager(wr->kernel_cb);
+  hapiInvokeCallback(wr->kernel_cb);
 
   // inform QD that the kernel is complete
-  QdProcess(1);
+  hapiQdProcess(1);
 
   return NULL;
 }
@@ -547,13 +544,13 @@ static void* deviceToHostCallback(void* arg) {
 
   // invoke user callback
   if (wr->device_to_host_cb) {
-    CUDACallbackManager(wr->device_to_host_cb);
+    hapiInvokeCallback(wr->device_to_host_cb);
   }
 
   hapiWorkRequestCleanup(wr);
 
   // inform QD that device-to-host transfer is complete
-  QdProcess(1);
+  hapiQdProcess(1);
 
   return NULL;
 }
@@ -568,12 +565,12 @@ static void* lightCallback(void *arg) {
   void* cb = *((void**)conv_msg_tmp);
 
   // invoke user callback
-  if (cb != NULL) {
-    CUDACallbackManager(cb);
+  if (cb) {
+    hapiInvokeCallback(cb);
   }
 
   // notify process to QD
-  QdProcess(1);
+  hapiQdProcess(1);
 
   return NULL;
 }
@@ -675,7 +672,7 @@ void hapiEnqueue(hapiWorkRequest* wr) {
   if (wr->host_to_device_cb) {
     // while there is an ongoing workrequest, quiescence should not be detected
     // even if all PEs seem idle
-    QdCreate(1);
+    hapiQdCreate(1);
 
 #ifdef HAPI_CUDA_CALLBACK
     addCallback(wr, AfterHostToDevice);
@@ -689,7 +686,7 @@ void hapiEnqueue(hapiWorkRequest* wr) {
 
   // add kernel callback
   if (wr->kernel_cb) {
-    QdCreate(1);
+    hapiQdCreate(1);
 
 #ifdef HAPI_CUDA_CALLBACK
     addCallback(wr, AfterKernel);
@@ -702,7 +699,7 @@ void hapiEnqueue(hapiWorkRequest* wr) {
   CsvAccess(gpu_manager).deviceToHostTransfer(wr);
 
   // add device-to-host transfer callback
-  QdCreate(1);
+  hapiQdCreate(1);
 #ifdef HAPI_CUDA_CALLBACK
   // always invoked to free memory
   addCallback(wr, AfterDeviceToHost);
@@ -755,8 +752,8 @@ hapiWorkRequest::hapiWorkRequest() :
 #endif
   }
 
-static void createPool(int *nbuffers, int n_slots, CkVec<BufferPool> &pools);
-static void releasePool(CkVec<BufferPool> &pools);
+static void createPool(int *nbuffers, int n_slots, std::vector<BufferPool> &pools);
+static void releasePool(std::vector<BufferPool> &pools);
 
 // Initialization of HAPI functionalities.
 void initHybridAPI() {
@@ -867,10 +864,10 @@ static inline void profileWorkRequestEvent(hapiWorkRequest* wr,
     char type = wr->comp_type;
     char phase = wr->comp_phase;
 
-    CkVec<hapiRequestTimeInfo> &vec = wr->avg_times_[index][type];
-    if (vec.length() <= phase){
+    std::vector<hapiRequestTimeInfo> &vec = wr->avg_times_[index][type];
+    if (vec.size() <= phase){
       vec.growAtLeast(phase);
-      vec.length() = phase+1;
+      vec.size() = phase+1;
     }
     switch (event) {
       case DataSetup:
@@ -895,12 +892,11 @@ static inline void profileWorkRequestEvent(hapiWorkRequest* wr,
 // TODO list the alignment/fragmentation issues with either of two allocation schemes:
 // if single, large buffer is allocated for each subpool
 // if multiple, smaller buffers are allocated for each subpool
-static void createPool(int *n_buffers, int n_slots, CkVec<BufferPool> &pools){
-  CkVec<size_t>& mempool_boundaries = CsvAccess(gpu_manager).mempool_boundaries_;
+static void createPool(int *n_buffers, int n_slots, std::vector<BufferPool> &pools){
+  std::vector<size_t>& mempool_boundaries = CsvAccess(gpu_manager).mempool_boundaries_;
 
   // initialize pools
   pools.reserve(n_slots);
-  pools.length() = n_slots;
   for (int i = 0; i < n_slots; i++) {
     pools[i].size = mempool_boundaries[i];
     pools[i].head = NULL;
@@ -956,18 +952,18 @@ static void createPool(int *n_buffers, int n_slots, CkVec<BufferPool> &pools){
   }
 }
 
-static void releasePool(CkVec<BufferPool> &pools){
-  for (int i = 0; i < pools.length(); i++) {
+static void releasePool(std::vector<BufferPool> &pools){
+  for (int i = 0; i < pools.size(); i++) {
     BufferPoolHeader* hdr = pools[i].head;
     if (hdr != NULL) {
       hapiCheck(cudaFreeHost((void*)hdr));
     }
   }
-  pools.free();
+  pools.clear();
 }
 
 static int findPool(size_t size){
-  int boundary_array_len = CsvAccess(gpu_manager).mempool_boundaries_.length();
+  int boundary_array_len = CsvAccess(gpu_manager).mempool_boundaries_.size();
   if (size <= CsvAccess(gpu_manager).mempool_boundaries_[0]) {
     return 0;
   }
@@ -994,7 +990,7 @@ static int findPool(size_t size){
 
     return boundary_array_len;
   }
-  for (int i = 0; i < CsvAccess(gpu_manager).mempool_boundaries_.length()-1; i++) {
+  for (int i = 0; i < CsvAccess(gpu_manager).mempool_boundaries_.size()-1; i++) {
     if (CsvAccess(gpu_manager).mempool_boundaries_[i] < size &&
         size <= CsvAccess(gpu_manager).mempool_boundaries_[i+1]) {
       return (i + 1);
@@ -1006,7 +1002,7 @@ static int findPool(size_t size){
 static void* getBufferFromPool(int pool, size_t size){
   BufferPoolHeader* ret;
 
-  if (pool < 0 || pool >= CsvAccess(gpu_manager).mempool_free_bufs_.length()) {
+  if (pool < 0 || pool >= CsvAccess(gpu_manager).mempool_free_bufs_.size()) {
     printf("[HAPI (%d)] getBufferFromPool, pool: %d, size: %zu invalid pool\n",
            CmiMyPe(), pool, size);
 #ifdef HAPI_MEMPOOL_DEBUG
@@ -1133,10 +1129,10 @@ void hapiPoolFree(void* ptr) {
 #ifdef HAPI_INSTRUMENT_WRS
 void hapiInitInstrument(int n_chares, char n_types) {
   avg_times_.reserve(n_chares);
-  avg_times_.length() = n_chares;
+  avg_times_.size() = n_chares;
   for (int i = 0; i < n_chares; i++) {
     avg_times_[i].reserve(n_types);
-    avg_times_[i].length() = n_types;
+    avg_times_[i].size() = n_types;
   }
   init_instr_ = true;
 }
@@ -1146,7 +1142,7 @@ static bool initializedInstrument() {
 }
 
 hapiRequestTimeInfo* hapiQueryInstrument(int chare, char type, char phase) {
-  if (phase < avg_times_[chare][type].length()) {
+  if (phase < avg_times_[chare][type].size()) {
     return &avg_times_[chare][type][phase];
   }
   else {
@@ -1155,19 +1151,19 @@ hapiRequestTimeInfo* hapiQueryInstrument(int chare, char type, char phase) {
 }
 
 void hapiClearInstrument() {
-  for (int chare = 0; chare < avg_times_.length(); chare++) {
-    for (int type = 0; type < avg_times_[chare].length(); type++) {
-      for (int phase = 0; phase < avg_times_[chare][type].length(); phase++) {
+  for (int chare = 0; chare < avg_times_.size(); chare++) {
+    for (int type = 0; type < avg_times_[chare].size(); type++) {
+      for (int phase = 0; phase < avg_times_[chare][type].size(); phase++) {
         avg_times_[chare][type][phase].transferTime = 0.0;
         avg_times_[chare][type][phase].kernelTime = 0.0;
         avg_times_[chare][type][phase].cleanupTime = 0.0;
         avg_times_[chare][type][phase].n = 0;
       }
-      avg_times_[chare][type].length() = 0;
+      avg_times_[chare][type].size() = 0;
     }
-    avg_times_[chare].length() = 0;
+    avg_times_[chare].size() = 0;
   }
-  avg_times_.length() = 0;
+  avg_times_.size() = 0;
   init_instr_ = false;
 }
 #endif // HAPI_INSTRUMENT_WRS
@@ -1184,7 +1180,7 @@ void hapiPollEvents() {
     if (cudaEventQuery(hev.event) == cudaSuccess) {
       // invoke Charm++ callback if one was given
       if (hev.cb) {
-        ((CkCallback*)hev.cb)->send(hev.cb_msg);
+        hapiInvokeCallback(hev.cb, hev.cb_msg);
       }
 
       // clean up hapiWorkRequest
@@ -1196,7 +1192,7 @@ void hapiPollEvents() {
       CpvAccess(n_hapi_events)--;
 
       // inform QD that an event was processed
-      QdProcess(1);
+      hapiQdProcess(1);
     }
     else {
       // stop going through the queue once we encounter a non-successful event
@@ -1267,7 +1263,7 @@ void hapiAddCallback(cudaStream_t stream, void* cb, void* cb_msg) {
 
   // while there is an ongoing workrequest, quiescence should not be detected
   // even if all PEs seem idle
-  QdCreate(1);
+  hapiQdCreate(1);
 }
 
 cudaError_t hapiMalloc(void** devPtr, size_t size) {

--- a/src/ck-core/ck.C
+++ b/src/ck-core/ck.C
@@ -437,13 +437,11 @@ void CkSectionID::pup(PUP::er &p) {
 /**** Tiny random API routines */
 
 #if CMK_CUDA
-void CUDACallbackManager(void *fn) {
-  if (fn != NULL) {
-    CkCallback *cb = (CkCallback*) fn;
-    cb->send();
+void CUDACallbackManager(void *fn, void *msg) {
+  if (fn) {
+    ((CkCallback*)fn)->send(msg);
   }
 }
-
 #endif
 
 void QdCreate(int n) {

--- a/src/ck-core/init.C
+++ b/src/ck-core/init.C
@@ -79,6 +79,14 @@ never be excluded...
 
 #if CMK_CUDA
 #include "hapi_impl.h"
+
+extern void (*hapiInvokeCallback)(void*, void*);
+extern void CUDACallbackManager(void*, void*);
+
+extern void (*hapiQdCreate)(int);
+extern void (*hapiQdProcess)(int);
+extern void QdCreate(int);
+extern void QdProcess(int);
 #endif
 
 void CkRestartMain(const char* dirname, CkArgMsg* args);
@@ -1691,6 +1699,9 @@ void _initCharm(int unused_argc, char **argv)
       CmiNodeBarrier();
     }
     hapiRegisterCallbacks();
+    hapiInvokeCallback = CUDACallbackManager;
+    hapiQdCreate = QdCreate;
+    hapiQdProcess = QdProcess;
 #endif
 
     if(CmiMyPe() == 0) {

